sklearn/preprocessing/_target_encoder.py

```diff
@@ -321,26 +380,142 @@ def _fit_encodings_all(self, X, y):
-                X_ordinal, y, n_categories, self.target_mean_, y_variance
+            encodings = _fit_encoding_fast_auto_smooth(
+                X_ordinal,
+                y,
+                n_categories,
+                target_mean,
+                y_variance,
             )
         else:
-            self.encodings_ = _fit_encoding_fast(
-                X_ordinal, y, n_categories, self.smooth, self.target_mean_
+            encodings = _fit_encoding_fast(
+                X_ordinal,
+                y,
+                n_categories,
+                self.smooth,
+                target_mean,
             )
 
-        return X_ordinal, X_known_mask, y, n_categories
+        if save_encodings:
+            self.encodings_ = encodings
+        return encodings
+
+    def _learn_multiclass_encodings(self, X_ordinal, y, n_categories, target_mean):
+        """Learn multiclass encodings.
+
+        Learn encodings for each class (c) then reorder encodings such that
+        the same features (f) are grouped together. `reorder_index` enables
+        converting from:
+        f0_c0, f1_c0, f2_c0, f0_c1, f1_c1, f2_c1
+        to:
+        f0_c0, f0_c1, f1_c0, f1_c1, f2_c0, f2_c1
+        """
+        n_features = self.n_features_in_
+        n_classes = len(self.classes_)
+
+        encodings = []
+        for i in range(n_classes):
+            y_class = y[:, i]
+            encoding = self._learn_encodings(
+                X_ordinal,
+                y_class,
+                n_categories,
+                target_mean[i],
+                save_encodings=False,
+            )
+            encodings += encoding
+
+        reorder_index = (
+            idx
+            for start in range(n_features)
+            for idx in range(start, (n_classes * n_features), n_features)
+        )
+        encodings = [encodings[idx] for idx in reorder_index]
+        return encodings
 
-    @staticmethod
     def _transform_X_ordinal(
-        X_out, X_ordinal, X_unknown_mask, indices, encodings, y_mean
+        self,
+        X_out,
+        X_ordinal,
+        X_unknown_mask,
+        indices,
+        encodings,
+        target_mean,
     ):
-        """Transform X_ordinal using encodings."""
-        for f_idx, encoding in enumerate(encodings):
-            X_out[indices, f_idx] = encoding[X_ordinal[indices, f_idx]]
-            X_out[X_unknown_mask[:, f_idx], f_idx] = y_mean
+        """Transform X_ordinal using encodings.
+
+        In the multiclass case, `X_ordinal` and `X_unknown_mask` have column
+        (axis=1) size `n_features`, while `encodings` has length of size
+        `n_features * n_classes`. `feat_idx` deals with this by repeating
+        feature indicies by `n_classes` E.g., for 3 features, 2 classes:
+        0,0,1,1,2,2
+
+        Additionally, `target_mean` is of shape (`n_classes`,) so `mean_idx`
+        cycles through 0 to `n_classes` - 1, `n_features` times.
+        """
+        for e_idx, encoding in enumerate(encodings):
```

Comment:
I prefer having the `if` statement outside the loop:

```python
if self.target_type_ == "multiclass":
    n_classes = len(self.classes_)
    ...
else:
    ...
```

I think this simpler to follow because there is no need to reason about a conditional during the loop.

Reply:
Thank you, good point.